examples/developer-hub-solidity/FAssetsSwapAndRedeem.sol

```diff
@@ -0,0 +1,211 @@
+    ISwapRouter public immutable router;
+    // FAssets assset manager interface
+    IAssetManager public immutable assetManager;
+    // FAssets token (FXRP)
+    IERC20 public immutable token;
+
+    // Path to swap WCFLR for FXRP
+    address[] public swapPath;
+
+    event RedemptionNeeded(uint256 amountIn, uint256 amountOut);
+    event SwapStarted(uint256 amount, address[] path, uint256 deadline);
+    event SwapCompleted(uint256[] amountsSent, uint256[] amountsRecv);
+    event RedeemStarted(uint256 lots, string redeemerUnderlyingAddressString);
+    event Redeemed(uint256 lots, uint256 redeemedAmountUBA);
+
+    constructor(
+        address _router,
+        address _assetManager,
+        address[] memory _swapPath
+    ) {
+        router = ISwapRouter(_router);
+        assetManager = IAssetManager(_assetManager);
+        swapPath = _swapPath;
+
+        token = IERC20(_swapPath[0]);
+    }
+
+    // 4. Main Function: `swapAndRedeem`
+
+    // Swap WCFLR for FXRP and redeem FAssets
+    // @param _lots: number of lots to redeem
+    // @param _redeemerUnderlyingAddressString: redeemer underlying address string (XRP address)
+    // @return amountOut: amount of FXRP received
+    // @return deadline: deadline of the swap
+    // @return amountsSent: amounts sent to the router
+    // @return amountsRecv: amounts received from the router
+    // @return _redeemedAmountUBA: amount of FAssets redeemed
+    function swapAndRedeem(
+        uint256 _lots,
+        string memory _redeemerUnderlyingAddressString
+    )
+        external
+        payable
+        returns (
+            uint256 amountOut,
+            uint256 deadline,
+            uint256[] memory amountsSent,
+            uint256[] memory amountsRecv,
+            uint256 _redeemedAmountUBA
+        )
+    {
+        // Calculate the amount needed to swap to FXRP and redeem
+        (uint256 _amountIn, uint256 _amountOut) = calculateRedemptionAmountIn(
+            _lots
+        );
+
+        require(
+            token.balanceOf(msg.sender) >= _amountIn,
+            "Insufficient token balance to swap"
+        );
+
+        // Check if the user has enough FXRP allowance
+        require(
+            token.allowance(msg.sender, address(this)) >= _amountIn,
+            "Insufficient FXRP allowance"
+        );
+
+        // Transfer tokens from msg.sender to this contract to swap to FXRP
+        require(
+            token.transferFrom(msg.sender, address(this), _amountIn),
+            "Transfer failed"
+        );
+
+        // Approve Uniswap router to spend the tokens
+        require(
+            token.approve(address(router), _amountIn),
+            "Router approval failed"
+        );
+
+        emit RedemptionNeeded(_amountIn, _amountOut);
+
+        // Set the deadline for the swap (10 minutes)
+        uint256 _deadline = block.timestamp + 10 minutes;
+        address[] memory path = swapPath;
+
+        emit SwapStarted(msg.value, swapPath, _deadline);
```

Comment:
Start/end emits are useless as there is no try/catch - so the whole transaction is atomic and either works or doesn't work